### DISCUSSION TITLE
Solving a linear system Ax = b where A, b are complex

### DISCUSSION THREAD

**Discussion by rtali-ia** [ginkgo-project/ginkgo#1577] [Q&A] [answered]
Hi There,

I am trying to solve a linear systems Ax = b where A,b are complex. I have saved A, b and x [the starting guess] as .mtx file. Here is my code

**using IndexType = int;
using vec = gko::matrix::Coo<std::complex<double>>;
using mtx = gko::matrix::Coo<std::complex<double>, IndexType>;

using cg = gko::solver::Cg<std::complex<double>>;

 auto A = gko::share(gko::read<mtx>(std::ifstream("../data/A.mtx"), exec));
 auto b = gko::read<mtx>(std::ifstream("../data/b.mtx"), exec);
 auto x = gko::read<mtx>(std::ifstream("../data/x0.mtx"), exec);

auto solver_gen =
        cg::build()
            .with_criteria(gko::stop::Iteration::build().with_max_iters(20u))
            .on(exec);

auto solver = solver_gen->generate(A);
solver->apply(b, x);**

I get the following error:

terminate called after throwing an instance of 'gko::DimensionMismatch'
  what():  /usr/local/include/ginkgo/core/base/lin_op.hpp:304: validate_application_parameters: attempting to combine operators this [62476 x 62476] and x [1 x 62476]: expected matching row length
Aborted

I am not sure where is the dimension mismatch coming from? How do I fix the code? 

Thanks

**Comment by upsj** [comment]
Your vector x0.mtx is stored as a row vector, it should be a column vector. You can either fix your output, or transpose it after reading:

```cpp
x = gko::as<vec>(x->transpose());
```

**Reply by rtali-ia** [reply]
Thank you. But I am still receiving the error

error: ‘class gko::matrix::Coo<std::complex<double>>’ has no member named ‘transpose’

Thanks for any inputs

**Reply by upsj** [reply]
Your vectors need to be of type `matrix::Dense`, not `matrix::Coo`

**Comment by rtali-ia** [comment]
Is there an example that I could use. I think I am struggling to deal with Complex vectors. Essentially, I pass a vector of size 62476 X 1 as a .mtx file to Ginkgo. This vector was created in python and converted to a .mtx file using **mmwrite** function. The problem is that when Ginkgo reads this .mtx file, it somehow converts itself into 1 X 62476 vector. I tried the instruction

`x = gko::as<vec>(x->transpose());`

but it still doesn't correct it. That is, it still prints the error

_/usr/local/include/ginkgo/core/base/lin_op.hpp:303: validate_application_parameters: attempting to combine operators this [62476 x 62476] and b [1 x 62476]: expected matching inner dimensions_

**Reply by MarcelKoch** [reply]
Could you provide us with your code? From the error you get, I would assume that either `x` or `x->transpose()` has to work.

**Reply by rtali-ia** [reply]
Thank you Marcel. Please find the full code [here](https://iastate.box.com/s/4pqgpa8ciaw095l10ip0sdioj52i5nev). I have put the full folder structure here including the _**CMakeLists.txt**_ . The main code is in _**simple-solver.cpp**_  All the data (**_A.mtx, b.mtx, x0.mtx_**) has been placed in the _**data**_ folder. You just need to follow the usual build protocol to reproduce the error. Thank you and much appreciate your eagerness to help me. Thank you again.

**Reply by MarcelKoch** [reply]
I don't have a box.com account. Please provide the code through some other means.

**Reply by rtali-ia** [reply]
No problem. Just created a public git repo. Thank you again. The repo is [here](https://github.com/rtali-ia/ginkgo-test)

**Reply by MarcelKoch** [reply, marked as answer]
Thank you. I was able to run your code without any issues after commenting out the transposing of both `x` and `b`.

**Reply by rtali-ia** [reply]
Thank you Marcel. Really appreciate you taking the time to look into this. I am getting excellent performance from Ginkgo after being able to run it. I am not even using the CUDA kernels. It seems adding the CUDA kernels will speed up the performance to several orders of magnitude. Thank you again for your patience.

**Reply by rtali-ia** [reply]
Also Marcel, I see that you have an experimental direct solver too. Is there a sample code that you could share with me that uses your direct solver?
